### PR TITLE
update password and update user

### PR DESCRIPTION
**Pull request by muhammadfaysal5**
i have implemented Update user update password api

### CODE REVIEW

File: controllers/userController.js [hunk 1]
```diff
@@ -175,17 +175,71 @@ const currentUser = async (req, res, next) => {
   ApiResponse.result(res, { user: data }, httpStatusCodes.OK);
 };
 
+const updateUser = async (req, res, next) => {
+  const { firstName, lastName, phone } = req.body;
+
+    const user = await Users.updateOne(
+      { _id: req.params.id },
+      { $set: {firstName, lastName, phone} }
```
Comment (hamzatariq7406) [inline controllers/userController.js]: what if I only want to update firstName not lastName ? In this case If I send only firstName in req.body lastName and phone will be set as undefined. Please update.

File: controllers/userController.js [hunk 2]
```diff
@@ -175,17 +175,71 @@ const currentUser = async (req, res, next) => {
   ApiResponse.result(res, { user: data }, httpStatusCodes.OK);
 };
 
+const updateUser = async (req, res, next) => {
+  const { firstName, lastName, phone } = req.body;
+
+    const user = await Users.updateOne(
+      { _id: req.params.id },
```
Comment (hamzatariq7406) [inline controllers/userController.js]: This should be object Id, not simple Id

File: controllers/userController.js [hunk 3]
```diff
@@ -175,17 +175,71 @@ const currentUser = async (req, res, next) => {
...
+};
+
 //logout user
 const logoutUser = (req, res, next) => {
   res.cookie("accessToken", "", { maxAge: 0 });
   ApiResponse.result(res, { status: "User logged out successfully" }, httpStatusCodes.OK);
 };
+const changePassword = async (req, res, next) => {
+  const { oldpassword, newpassword, confirmpassword } = req.body;
+   if (!oldpassword || !newpassword || !confirmpassword){
```
Comment (hamzatariq7406) [inline controllers/userController.js]: we don't send confirm password on backend. Its just a validation thing for frontend

File: controllers/userController.js [hunk 4]
```diff
@@ -175,17 +175,71 @@ const currentUser = async (req, res, next) => {
...
+    throw new ApiError(
+      httpStatusCodes.BAD_REQUEST,
+      "All fields are required",
+      httpStatusCodes.BAD_REQUEST
+    );
+   }
+   if (newpassword!==confirmpassword){
```
Comment (hamzatariq7406) [inline controllers/userController.js]: Remove this code, this validation handles only on client side not server side

File: controllers/userController.js [hunk 5]
```diff
@@ -175,17 +175,71 @@ const currentUser = async (req, res, next) => {
...
+    throw new ApiError(
+      httpStatusCodes.BAD_REQUEST,
+      "New password and confirm password is not matched",
+      httpStatusCodes.BAD_REQUEST
+    );
+}
+   const user = await Users.findOne({ _id: req.user._userInfo });
```
Comment (hamzatariq7406) [inline controllers/userController.js]: Extract Id from userInfor and pass it as object Id

File: controllers/userController.js [hunk 6]
```diff
@@ -175,17 +175,71 @@ const currentUser = async (req, res, next) => {
...
+   if(user && (await bcrypt.compare(oldpassword, user.password))){
+    const salt = await bcrypt.genSalt(10);
+  const hashedPassword = await bcrypt.hash(newpassword, salt);
+
+    await Users.updateOne(
+      { _id: req.user._userInfo },
```
Comment (hamzatariq7406) [inline controllers/userController.js]: Same, needs to be object Id and extracted inside _userInfo

File: controllers/userController.js [hunk 7]
```diff
@@ -175,17 +176,120 @@ const currentUser = async (req, res, next) => {
...
+  const objectId = req.user._userInfo
+   if (!oldpassword || !newpassword ){
+    throw new ApiError(
+      httpStatusCodes.BAD_REQUEST,
+      "All fields are required",
+      httpStatusCodes.BAD_REQUEST
+    );
+   }
+
+   console.log('user info ', objectId)
```
Comment (hamzatariq7406) [inline controllers/userController.js]: Remove console

File: controllers/userController.js [hunk 8]
```diff
@@ -175,17 +176,120 @@ const currentUser = async (req, res, next) => {
...
   res.cookie("accessToken", "", { maxAge: 0 });
   ApiResponse.result(res, { status: "User logged out successfully" }, httpStatusCodes.OK);
 };
+const changePassword = async (req, res, next) => {
+  const { oldpassword, newpassword } = req.body;
+  const objectId = req.user._userInfo
+   if (!oldpassword || !newpassword ){
+    throw new ApiError(
+      httpStatusCodes.BAD_REQUEST,
+      "All fields are required",
```
Comment (hamzatariq7406) [inline controllers/userController.js]: Message should be: "Old Password or new password is not provided"

File: controllers/userController.js [hunk 9]
```diff
@@ -175,17 +176,120 @@ const currentUser = async (req, res, next) => {
...
+      httpStatusCodes.BAD_REQUEST
+    );
+   }
+
+   console.log('user info ', objectId)
+   
+   const user = await Users.findOne({ _id: objectId });
```
Comment (hamzatariq7406) [inline controllers/userController.js]: what is objectId here ? Can you see from my previous implementation find on basis of _id ? or you can search online

File: controllers/userController.js [hunk 10]
```diff
@@ -175,17 +176,120 @@ const currentUser = async (req, res, next) => {
...
+   if(user && (await bcrypt.compare(oldpassword, user.password))){
+    const salt = await bcrypt.genSalt(10);
+  const hashedPassword = await bcrypt.hash(newpassword, salt);
+
+    await Users.updateOne(
+      { _id: objectId },
```
Comment (hamzatariq7406) [inline controllers/userController.js:220]: Same issue here

**Review by hamzatariq7406 — CHANGES_REQUESTED** [review]

File: controllers/userController.js (hunk 9 above)
Comment (muhammadfaysal5) [reply controllers/userController.js]: i have extracted objectId from _userInfo
const objectId = req.user._userInfo

Comment (hamzatariq7406) [reply controllers/userController.js]: No, do it like this:
new Types.ObjectId(req.user._userInfo.id)

import is like this
import { Types } from "mongoose";

**Review by hamzatariq7406 — CHANGES_REQUESTED** [review]

**Pull request merged**